### PR TITLE
Compute value hashes in parallel for v2Merkle

### PR DESCRIPTION
**Pull request by dartdart26**
Spawn a separate thread to compute value hashes while updating the
sparse merkle tree. Idea is to have the current value hash ready for the
update while the separate thread is computing another one.

Implement by using std::promise/future pairs. Introduce the ValueHashes
type to encapsulate the mapping between keys, values, promises and
futures.

Assume the same order for subsequent iterations of SetOfKeyValuePairs
objects, provided that we don't mutate them in any way. That, combined
with the assumption that the tree update takes longer than value hash
calculation, should ensure that this change is actually an optimization.
Note: The documentation for std::unordered_map (SetOfKeyValuePairs)
states that read-only operations do not invalidate iterators. Therefore,
we can expect that subsequent iterations will have the same order.

Benchmark data on an i7-9750H CPU @ 2.60GHz:
Before:
```
-----------------------------------------------------------------------
Benchmark                                Time          CPU   Iterations
-----------------------------------------------------------------------
Blockchain/addBlock/128/4/1024    4456033 ns      4456167 ns        154
Blockchain/addBlock/512/32/1024   24865546 ns     24865500 ns        29
Blockchain/addBlock/2048/8/4096  159640857 ns    159627593 ns         4

```
After:
```
-----------------------------------------------------------------------
Benchmark                                Time          CPU   Iterations
-----------------------------------------------------------------------
Blockchain/addBlock/128/4/1024    4051238 ns      3924905 ns        175
Blockchain/addBlock/512/32/1024   22804525 ns     22655801 ns        32
Blockchain/addBlock/2048/8/4096  119120398 ns    118909535 ns         6
```

Remove unused <iostream> include and 'using namespace std' directive.

### CODE REVIEW

File: kvbc/src/sparse_merkle/tree.cpp [hunk 1]
```diff
@@ -13,14 +13,58 @@
...
+#include <iterator>
+#include <unordered_map>
 
 using namespace concordUtils;
 
 namespace concord::kvbc::sparse_merkle {
 using namespace detail;
 
+// Computes value hashes in a separate thread.
+class ValueHashes {
```
Comment (teoparvanov) [inline kvbc/src/sparse_merkle/tree.cpp:26]: AsyncValueHashes ?

Comment (dartdart26) [reply kvbc/src/sparse_merkle/tree.cpp:26]: Yep, will rename.

File: kvbc/src/sparse_merkle/tree.cpp [hunk 2]
```diff
@@ -141,11 +185,17 @@ UpdateBatch Tree::update_impl(const concord::kvbc::SetOfKeyValuePairs& updates,
...
+  // should ensure that this is actually an optimization.
+  // Note: The documentation for std::unordered_map (SetOfKeyValuePairs) states that read-only operations do not
+  // invalidate iterators. Therefore, we can expect that subsequent iterations will have the same order.
+  auto value_hashes = ValueHashes(updates);
   for (auto&& [key, val] : updates) {
-    auto leaf_hash = hasher.hash(val.data(), val.length());
     LeafNode leaf_node{val};
     LeafKey leaf_key{hasher.hash(key.data(), key.length()), version};
-    LeafChild child{leaf_hash, leaf_key};
+    LeafChild child{value_hashes[key], leaf_key};
```
Comment (andrewjstone) [inline kvbc/src/sparse_merkle/tree.cpp:198]: Doesn't this block, since get is called on the future?

Comment (dartdart26) [reply kvbc/src/sparse_merkle/tree.cpp:198]: You are correct, it does. Assumption here is that value hash calculation is faster than the tree update. As discussed offline, I am closing this PR for now. Will open another PR that implements an std::async()-like thread pool as an utility. Will further measure performance with it and decide if we should calculate value hashes in parallel and, if yes, how many threads should be used.

**Comment by dartdart26** [comment]
Closing this PR as per: https://github.com/vmware/concord-bft/pull/670#discussion_r442877274

**Pull request closed without merging**